task-maker-async/src/worker.rs

```diff
@@ -0,0 +1,566 @@
+        config.stack_limit(stack * 1024);
+    }
+    let multiproc = Some(1) != execution.limits.nproc;
+    config.syscall_filter(SyscallFilter::build(
+        multiproc,
+        !execution.constraints.read_only,
+    ));
+    // has to be writable for mounting stuff in it
+    config.mount(boxdir.join("etc"), "/etc", true);
+
+    for dir in READABLE_DIRS {
+        if Path::new(dir).is_dir() {
+            config.mount(dir, dir, false);
+        }
+    }
+    for dir in &execution.constraints.extra_readable_dirs {
+        if dir.is_dir() {
+            config.mount(dir, dir, false);
+        }
+    }
+    if execution.constraints.mount_tmpfs {
+        config.mount_tmpfs(true);
+    }
+    if execution.constraints.mount_proc {
+        config.mount_proc(true);
+    }
+    match &execution.command {
+        ExecutionCommand::System(cmd) => {
+            if let Ok(cmd) = which::which(cmd) {
+                config.executable(cmd);
+            } else {
+                bail!("Executable {:?} not found", cmd);
+            }
+        }
+        ExecutionCommand::Local(cmd) => {
+            let host_cmd = boxdir.join("box").join(cmd);
+            validate_local_executable(&host_cmd)?;
+            config.executable(boxroot.join(cmd));
+        }
+    };
+    for arg in execution.args.iter() {
+        config.arg(arg);
+    }
+    // drop root privileges in the sandbox
+    // TODO(veluca): is 1000 always the correct UID?
+    config.uid(1000);
+    config.gid(1000);
+    Ok(config)
+}
+
+struct ExecutionInfo {
+    dir: PathBuf,
+    execution: Execution,
+    sandbox_config: SandboxConfiguration,
+}
+
+async fn prepare_one_execution(
+    store: StoreClient,
+    execution: Execution,
+    group_dir: PathBuf,
+) -> Result<ExecutionInfo, Error> {
+    // TODO(veluca): this requires execution names to be valid directory names.
+    let dir = group_dir.join(&execution.name);
+
+    let path_for_file = Mutex::new(HashMap::new());
+
+    // Fetch all the input files and link all the FIFOs.
+    try_join_all(execution.files.iter().map(|(path, file)| async {
+        let local_path = box_path(&dir, &path.clone());
+        if path_for_file
+            .lock()
+            .unwrap()
+            .insert(path.clone(), local_path.clone())
+            .is_some()
+        {
+            return Err(anyhow::anyhow!("Duplicate file specified"));
+        }
+        match file.clone() {
+            ExecutionFileMode::Input(input) => {
+                write_to_sandbox(&store, &input, &local_path).await?
+            }
+            ExecutionFileMode::Fifo(fifo) => {
+                hard_link(&group_dir.join(FIFO_DIR_NAME).join(fifo), local_path).await?;
+            }
+            ExecutionFileMode::Output => {}
+        };
+        Ok(())
+    }))
+    .await?;
+
+    // Write /etc/passwd in the sandbox root, as some software expects it to be present.
+    // TODO(veluca): this also ought to be done by tabox.
```

Comment:
I see your point, but OTOH the fact that /etc/passwd is necessary is pretty much an implementation detail of the fact that tabox chroots...

Reply:
Since it's the sandbox to create the root filesystem and bind-mount the other directory into it, it makes sense that it's responsibility of the sandbox to provide a minimal set of commonly expected configuration files to guarantee a proper operation. Of course if the `/etc` of the sandbox is mounted by the application that uses it, these file will override the default generated ones (i.e. the /etc directory is generated in the tmpdir and then bind mounts are applied to it).